ml-agents/mlagents/trainers/torch/networks.py

```diff
@@ -41,17 +42,47 @@ def __init__(
             else 0
         )
 
-        self.processors, self.embedding_sizes = ModelUtils.create_input_processors(
+        self.processors, self.embedding_sizes, var_len_indices = ModelUtils.create_input_processors(
             observation_specs,
             self.h_size,
             network_settings.vis_encode_type,
             normalize=self.normalize,
         )
 
-        total_enc_size = sum(self.embedding_sizes) + encoded_act_size
-        self.linear_encoder = LinearEncoder(
-            total_enc_size, network_settings.num_layers, self.h_size
-        )
+        if len(var_len_indices) > 0:
+            # There are some variable length observations and they need to be processed separately
+            x_self_len = sum(self.embedding_sizes)  # The size of the "self" embedding
+            entities_sizes = [
+                observation_specs[idx].shape[1] for idx in var_len_indices
+            ]
+            entities_max_len = [
+                observation_specs[idx].shape[0] for idx in var_len_indices
+            ]
+
+            self.x_self_encoder = LinearEncoder(x_self_len, 2, self.h_size // 2)
+            self.var_len_encoders = torch.nn.ModuleList(
+                [
+                    LinearEncoder(ent_size, 2, self.h_size // 2)
+                    for ent_size in entities_sizes
+                ]
+            )
+            self.entities_embeddings = EntityEmbeddings(
+                self.h_size // 2,
+                [self.h_size // 2] * len(var_len_indices),
+                entities_max_len,
+                self.h_size,
+            )
+            self.rsa = ResidualSelfAttention(self.h_size, sum(entities_max_len))
+
+            total_enc_size = self.h_size // 2 + self.h_size
+
+            n_layers = max(1, network_settings.num_layers - 2)
+        else:
+            total_enc_size = sum(self.embedding_sizes)
```

Comment:
This line wont go off if there is a variable length obs. If there's a separate vector obs, total_enc_size won't include that.